File: solutions/typescript/collatz-conjecture/1/collatz-conjecture.ts
export function steps(num: number): number | never {
  if (num < 1 || !Number.isInteger(num)) {
    throw new Error('Only positive integers are allowed')
  }
  let steps = 0
  let cur = num
  while (cur !== 1) {
    ++steps
    if ((cur & 1) === 0) {
      cur >>= 1
    } else {
      cur += (cur << 1) + 1
    }
  }
  return steps
}

File: solutions/python/collatz-conjecture/1/collatz_conjecture.py
def steps(number):
    if number <= 0:
        raise ValueError("Only positive integers are allowed")
    step = 0
    while number != 1:
        step += 1
        if number % 2 == 0:
            number //= 2
        else:
            number *= 3
            number += 1
    return step